### PR TITLE
fix: explicit source dependency is not satisfied by direct origin

### PR DESCRIPTION
**Pull request by ralbertazzi**
Resolves: #7722

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

cc @dimbleby

### CODE REVIEW

**Comment by ralbertazzi** [comment]
There seems to be a weird issue with pytest fixtures in the failing test 🤔

File: tests/puzzle/test_provider.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def test_search_for(
             Dependency("foo", ">=2"),
             URLDependency("foo", SOME_URL),
             [Package("foo", "3")],
-            [],
+            [Package("foo", "3")],
```
Comment (ralbertazzi) [inline tests/puzzle/test_provider.py:116]: Looks like correct behaviour to me: the dependency is not satisfied by the URL package and so it falls back to checking the repository (which exactly the proposed change)

**Comment by dimbleby** [comment]
So far as the code change goes I'm basically reviewing my own suggestion here.  Looks fine to me, but this is not an independent review.

I think I would have expected to write a testcase here that more closely follows the actual problem as laid out in the issue report.  Probably something in `test_installer.py` that sets things up and runs `poetry install` and checks that the expected thing happens.

I always find those testcases unpleasant to write, which is partly what put me off doing anything with this one in the first place.  If you can convince yourself - and someone with the commit bit - that what you have is sufficient, then your life will be easier.

**Comment by ralbertazzi** [comment]
I added a test, LMK if that's what you were looking for ;)

**Review by radoering — APPROVED** [review]

**Pull request merged**